Design updates to the Publish popover (`DateTimePicker`)

noisysocks opened the pull request:
## What?
Implements the new design for the Publish popover (`DateTimePicker`) that's in https://github.com/WordPress/gutenberg/issues/39082.

Since we're changing most of the CSS anyway, I've taken this opportunity to convert most of `DateTimePicker` to use existing WordPress components and Emotion styling.

## Why?
Part of https://github.com/WordPress/gutenberg/issues/39082.

## How?
- Updates the design of `DateTimePicker`, `DatePicker` and `TimePicker`.
- Changes `DateTimePicker`, `DatePicker` and `TimePicker` to use existing WordPress components and Emotion styling where it makes sense to do so.
  - I kept all of the `react-dates` theming as CSS. Emotion doesn't really add much here since we have to use CSS selectors anyway.
- Re-organises `DateTimePicker` folder structure to be more consistent with newer components.
  - Each sub-component in `date-time` now has its own folder.
- Allows removal of _Reset_ button and _Help_ functionality from `DateTimePicker`.
  - The new design does not have this.
  - This is implemented using a `__next` prop. We should eventually make this the default.
- Adds `PublishDateTimePicker` to block-editor.
  - This exposes a specialised `DateTimePicker` that is used for changing a _post_ publish date.
  - Basically it just adds the _Publish_ heading.

## Testing Instructions
1. Create a post.
2. Open the post settings sidebar.
3. Change the _Publish_ date.4. 

Also play with the `DateTimePicker` storybook in `npm run storybook:dev`.

There are existing unit tests for `DateTimePicker` which hopefully I didn't break.

## Screenshots or screencast
| English | Arabic |
| - | - |
| <img width="280" alt="Screen Shot 2022-05-19 at 18 03 10" src="https://user-images.githubusercontent.com/612155/169244006-2eee5bc4-eaf5-4781-b17a-120c85edf8d6.png"> | <img width="279" alt="Screen Shot 2022-05-19 at 18 02 22" src="https://user-images.githubusercontent.com/612155/169244041-c169b8ee-783e-4d8b-88aa-934d4eb9cb22.png"> |

noisysocks commented on packages/block-editor/src/components/publish-date-time-picker/index.js:
```diff
@@ -0,0 +1,43 @@
+} ) {
+	return (
+		<div className="block-editor-publish-date-time-picker">
+			{ /* TODO: This header is essentially the same as the one in <PostVisiblity />. DRY it up. */ }
```
I'll do this in a follow-up as it warrants some discussion. Also this PR is big enough 😀

noisysocks commented on packages/components/src/date-time/date/index.tsx:
```diff
@@ -2,30 +2,25 @@
 import type { Moment } from 'moment';
 import { noop } from 'lodash';
-
-// `react-dates` doesn't tree-shake correctly, so we import from the individual
```
Didn't mean to delete this comment.

noisysocks commented on packages/components/src/date-time/date/index.tsx:
```diff
@@ -2,30 +2,25 @@
 import type { Moment } from 'moment';
 import { noop } from 'lodash';
-
-// `react-dates` doesn't tree-shake correctly, so we import from the individual
-// component here.
```
Oops, don't delete these two lines.

noisysocks commented on packages/components/src/date-time/date/index.tsx:
```diff
@@ -69,14 +64,13 @@ function DatePickerDay( { day, events = [] }: DatePickerDayProps ) {
-			className={ classnames( 'components-datetime__date__day', {
-				'has-events': events?.length,
-			} ) }
+			className="components-datetime__date__day"
```
We're no longer using most of the class names but I've tried to keep them for backwards compatibility. I don't think this is a hard requirement, but maybe nice to do anyway.

ciampo replied:
Yes, we should keep all of those legacy classnames around for compat reasons. Maybe we should add a comment about that, so that it doesn't get considered "dead code" and gets removed in the future.

noisysocks commented:
I've still got some tasks remaining before we can merge this (see description) but I'm happy with the broad strokes and think this is ready for a first round of review.

andrewserong commented:
Thanks for the ping @noisysocks, I think I'll run out of time today to take a closer look, but I'm looking forward to getting stuck into review 🤓

At first glance, the design's looking really nice in the editor 👍

<img src="https://user-images.githubusercontent.com/14988353/168972018-678bed68-c6e4-420e-8b7c-00dbae22275f.png" width="250" />

One thing I noticed is that if you click on the AM / PM buttons it closes the popover, unlike the other buttons in the popover. Not sure if focus is being moved somewhere?

andrewserong commented on packages/components/src/date-time/time/index.tsx:
```diff
@@ -0,0 +1,309 @@
+			max={ 31 }
+			hideHTMLArrows
+			isPressEnterToChange
+			isDragEnabled={ false }
```
Just wondering why we're disabling the drag feature? We don't currently have this behaviour on `trunk`, but I wondered if it'd make it quicker for Day and Time to be able to click and drag it up and down?

noisysocks replied:
It's not a great UX, unfortunately, because editing the post date (using the `editPost()` action) is fairly slow and re-rendering the `react-dates` calendar component is _very_ slow. The drag gesture ends up feeling really sluggish. We could spend some time optimising all of that but I don't really think it's worth the investment for now.

andrewserong replied:
Good call... perhaps an exploration for an enhancement at some point after this lands!

noisysocks commented:
> One thing I noticed is that if you click on the AM / PM buttons it closes the popover, unlike the other buttons in the popover. Not sure if focus is being moved somewhere?

👍 I made it so that the popover remains open. The alternative is that we auto-close it after a date is selected but I don't like that because then it makes it hard to select a date **and** a time. Also we have an X now.

javierarce commented:
This is looking great! I just have just one comment:

When you have scheduled another post we show a small blue dot under the day number. But if you hover over the day or select it, the dot disappears. I propose changing the color of the dot so it's always visible. That would mean white for when the day is selected or selected and hovered. 

https://user-images.githubusercontent.com/4933/169019189-7e6ee1f2-7e64-4a1d-96e4-9f084f75c6ba.mov


Then I noticed a weird reload the first time you select a day from a different month. I guess this is not related to this PR, but I just wanted to mention it anyway in case it's helpful. 

https://user-images.githubusercontent.com/4933/169019483-191eb3a4-c15b-4829-81b0-935f231670df.mov

andrewserong commented on packages/components/src/number-control/index.js:
```diff
@@ -43,9 +44,10 @@ export function NumberControl(
+		const number = isStepAny
 			? Math.min( max, Math.max( min, value ) )
 			: roundClamp( value, min, max, stepOverride ?? baseStep );
+		return number.toString().padStart( pad, '0' );
```
The `0` padding looks like a useful feature!

From a quick glance, it looks like this changes the implicit return type for `constrainValue` from a number to a string. Would it be worth splitting out this change (and the change to `SelectControl`) into a separate PR(s)? Since NumberControl is fairly widely used, and can receive a state reducer, it might be worth us pinging a few folks on the change in isolation to get a few more eyes on the potential implications (I've been tripped up trying to make changes to NumberControl before 😄).

andrewserong replied:
Now that I think of it, the conversion to `string` might not be such a problem (and could be desired). I see from @ciampo's earlier NumberControl PR to refactor to TypeScript (#38753) it proposes introducing an `ensureString` function that gets used in the reducer, so the change here might be useful for that, too.

noisysocks replied:
Sure, I can split this out!

Changing the return type to string is intentional. If you look at the typings for the reducer state in `SelectControl` you'll see that `value` is supposed to be a string.

andrewserong replied:
Great, thanks for confirming!

ciampo replied:
I took `NumberControl` for a spin and played around with the new `pad` property, and found a few instances where it behaved unexpectedly (for example, the `0` padding is not added when typing a value and blurring the input field).

https://user-images.githubusercontent.com/1083581/169321627-1e57b64c-7a1c-4162-9826-4aa8bb83b847.mp4

Considering that `InputControl` (and consequently `NumberControl`) [is undergoing a few changes / fixes](https://github.com/WordPress/gutenberg/pull/40518), and that `NumberControl` isn't typed yet, my recommendation for now is to add this feature directly to `TimePicker` if possible (and consider porting it to `NumberControl` once those fixes/improvements are applied).

andrewserong replied:
> my recommendation for now is to add this feature directly to TimePicker if possible (and consider porting it to `NumberControl` once those fixes/improvements are applied).

That sounds like a good idea to me, too.

noisysocks commented:
> When you have scheduled another post we show a small blue dot under the day number. But if you hover over the day or select it, the dot disappears. I propose changing the color of the dot so it's always visible. That would mean white for when the day is selected or selected and hovered.

Good thinking! I've done this.

> Then I noticed a weird reload the first time you select a day from a different month. I guess this is not related to this PR, but I just wanted to mention it anyway in case it's helpful.

Let's look into this separately. It happens because of https://github.com/react-dates/react-dates/issues/1320 which is fixed but not yet released.

noisysocks commented:
OK, happy with this now 🙂 I fixed unit tests, tweaked some of the margins, fixed some issues with RTL languages, and tested everything in multiple browsers.

I'll split out some of the `@wordpress/components` stuff into a separate PR for ease of review tomorrow, but don't let that stop you from giving this a whirl!

ciampo commented on packages/block-editor/src/components/publish-date-time-picker/style.scss:
```diff
@@ -0,0 +1,20 @@
+	text-decoration: underline;
+}
+
+.block-editor-publish-date-time-picker__close.has-icon {
```
I wonder if there's a way to apply these styles that doesn't involve using the `has-icon` classname, which should be considered as an implementation detail of the `Button` component

noisysocks replied:
Do you think the `&&&` hack would be better? It's just specificity that we need.

ciampo replied:
I think so — the `&&&` hack is more upfront about why is being used (specificity) and would allow us to avoid using hardcoded internal components classnames :)

noisysocks replied:
Hm, doesn't look like `sass` lets you do the `&&&` hack. I think how it is currently is the best we can do.

```
Error: "&" may only used at the beginning of a compound selector.
```

ciampo replied:
Off the top of my head, I remember being able to do something like that by using [interpolation](https://sass-lang.com/documentation/interpolation) — something like

```sass
  #{ & }#{ & }#{ & } {
    // styles
  }
```

although I'm not 100% sure

stokesman replied:
The hack I usually reach for is attribute selectors.
```suggestion
[class].block-editor-publish-date-time-picker__close {
```
You can repeat the same one as required for specificity but here you might only need the single.

ciampo commented on packages/components/src/select-control/types.ts:
```diff
@@ -20,7 +20,7 @@ export interface SelectControlProps
 			| 'suffix'
 		>,
-		Pick< BaseControlProps, 'help' > {
+		Pick< BaseControlProps, 'help' | '__nextHasNoMarginBottom' > {
```
While we're slowly moving towards using Storybook as the main reference for props docs, we should still update the component's README for now — could we add an entry for the `__nextHasNoMarginBottom` prop? (we could copy the text from `BaseControl`'s README)

ciampo commented on packages/components/package.json:
```diff
@@ -62,7 +62,7 @@
 		"lodash": "^4.17.21",
 		"memize": "^1.1.0",
-		"moment": "^2.22.1",
+		"moment": "^2.26.0",
```
Is there a particular reason why we're updating `moment`, or is it just to keep it up to date?

noisysocks replied:
Yeah, this was a little dependency nightmare. I had to update `@types/react-dates` because some functions were missing from the older type definitions (`renderMonthElement`, I think) and then that newer version of the typings required `moment` to be updated because it expected to see more functions in the `moment.Moment` interface than what existed. It's all very silly since none of it actually makes a difference to the runtime code that's generated.

ciampo replied:
Got it, thank you!

ciampo commented on packages/components/src/date-time/date/index.tsx:
```diff
@@ -2,30 +2,25 @@
-import UntypedDayPickerSingleDateController from 'react-dates/lib/components/DayPickerSingleDateController';
-import type { DayPickerSingleDateController } from 'react-dates';
-const TypedDayPickerSingleDateController = UntypedDayPickerSingleDateController as DayPickerSingleDateController;
+import DayPickerSingleDateController from 'react-dates/lib/components/DayPickerSingleDateController';
```
Is this fixed by updating `@types/react-dates` to its latest version?

Neat!

noisysocks replied:
Yes! A happy accident.

ciampo commented on packages/components/src/date-time/style.scss:
```diff
@@ -1,262 +1 @@
-.components-popover .components-datetime__date {
-	padding-left: $grid-unit-05;
-}
+@import "./date/style.scss";
```
To recap:

- `date-time/date/datepicker.scss` is basically a copy-paste of the original styles in `react-dates`, plus a few tweaks
- `date-time/date/style.scss` overrides/adds theming to those styles, using the classnames from `react-dates`
- `date-time/styles.scss` just imports the styles mentioned above
- `date-time/styles.ts` contains all remaining styles for the components written with Emotion


In case that's correct, to make things a bit more simple/clear we could:

- rename `date-time/date/datepicker.scss` to `date-time/date/lib.scss`
- move the code in `date-time/date/style.scss` to `date-time/styles.scss` (which would import `./date/lib.scss`)
- keep `date-time/styles.ts` as-is

Not a strong preference at all, though — WDYT?

noisysocks replied:
> * `date-time/date/datepicker.scss` is basically a copy-paste of the original styles in `react-dates`

Correct.

> * plus a few tweaks

I don't actually think there are any tweaks. I'm not really sure why we vendor the stylesheet instead of using the one that's assumedly in `node_modules`.

> * `date-time/styles.scss` just imports the styles mentioned above

Correct.

> * `date-time/styles.ts` contains all remaining styles for the components written with Emotion

The `styles.ts` files are all within the component subfolders, so it's `date-time/date/styles.ts`.

---

> * rename `date-time/date/datepicker.scss` to `date-time/date/lib.scss`

Sounds good. I'll first quickly check to see we can just use what's in `node_modules`.

> * move the code in `date-time/date/style.scss` to `date-time/styles.scss` (which would import `./date/lib.scss`)

No strong preference here. I suppose it's one less "go to definition" command you have to do when following the code, so yep sure!

ciampo replied:
Your plan sounds good to me!

> I don't actually think there are any tweaks. I'm not really sure why we vendor the stylesheet instead of using the one that's assumedly in node_modules.

The reason why I mentioned that is because of this comment at the start of the file: https://github.com/WordPress/gutenberg/blob/2aadd5fcea1c15dd6768886d8af635de5780b362/packages/components/src/date-time/date/datepicker.scss#L2-L3

noisysocks replied:
OK looked more into this and there's some history on why we have a vendored copy of `datepicker.scss`: https://github.com/WordPress/gutenberg/pull/26534. Basically it uses `darken()` for some reason and so needs to have the `.scss` extension. Vendoring it is the easiest way to do so.

I also am thinking it's still best to keep both `date-time/date/style.scss` and `date-time/style.scss` as it makes it more explicit that **only** `DatePicker` uses `react-dates`.

Note to self: Really ought to investigate using something other than `react-dates` (maybe `react-calendar`).

ciampo replied:
Thank you for the context!

> Really ought to investigate using something other than `react-dates`

That would be great!

I personally would also love to find an alternative to `moment`, mostly to save on bundle size

ciampo commented on packages/components/src/date-time/time/styles.ts:
```diff
@@ -0,0 +1,119 @@
+import NumberControl from '../../number-control';
+import SelectControl from '../../select-control';
+
+export const Wrapper = styled.div``;
```
@sarayourfriend pardon the ping, but I kind of lost track on whether we have a strong preference for Styled components vs classnames + `cx` approach

noisysocks replied:
We have both patterns in the codebase so I was confused about this too. I landed on using `styled()` because I find the `useCx` / `cx` statements to be kind of distracting when reading the component's code. Not a strong opinion though!

ciampo commented on packages/components/src/number-control/test/index.js:
```diff
@@ -91,7 +91,7 @@ describe( 'NumberControl', () => {
 				expect( spy ).toHaveBeenCalledTimes( 2 );
 				expect( spy ).toHaveBeenNthCalledWith( 1, '1' );
-				expect( spy ).toHaveBeenNthCalledWith( 2, 4 );
+				expect( spy ).toHaveBeenNthCalledWith( 2, '4' );
```
The fact that we had to update a few of the expected values from numbers to strings also in `NumberControl`'s unit tests makes me want to stay on the side of caution here and delay these changes until we have some more clarity on how `NumberControl` handles its `value` internally (number vs string)

noisysocks replied:
Yep ok, I'll try to keep the changes in `TimePicker` for now.

I do think this is a mistake in `NumberControl`, though, and that `value` should always be a string. You can see from these tests that currently sometimes `onChange` is called with a number and sometimes with a string which is bananas! 🍌 It will come up when `NumberControl` is converted to TypeScript because we'll see errors about coercing `NumberControl`'s `value: number | string` to `InputControl`'s `value: string`. `NumberControl` is supposed to be a glorified `input[type=number]` so I think it makes sense that value should work the same as it does in `HTMLInputElement`.

ciampo replied:
I had already one attempt at typing `NumberControl`, and I had to give up mostly because of the ambiguity around which type should be used for `value` (`string`, like `HTMLInputElement`, or `number`, since it's an input field dealing with numbers?). We decided to punt that refactor forward, and instead refactor to TypeScript its dependent components: `UnitControl` (done) and `RangeControl` (WIP).

Once `RangeControl` is typed, I'll attempt the refactor once again, and I'll make sure to involve you and the rest of the folks helping this effort!

ciampo commented on packages/components/src/date-time/time/test/index.tsx:
```diff
@@ -1,15 +1,20 @@
 describe( 'TimePicker', () => {
-	it( 'should call onChange with updated date values', () => {
+	it( 'should call onChange with updated date values', async () => {
+		const user = userEvent.setup( {
```
Great to see `user-event` being used instead of `fireEvent`!

ciampo commented on packages/components/src/date-time/time/test/index.tsx:
```diff
@@ -89,13 +106,18 @@ describe( 'TimePicker', () => {
+		await user.keyboard( '{Enter}' );
 
-		expect( onChangeSpy ).not.toHaveBeenCalled();
+		expect( onChangeSpy ).toHaveBeenCalledWith( '1986-10-18T00:00:00' );
```
This change of behavior in the tests is interesting — any ideas why it happened?

I wonder whether it is a good change or if we should instead treat it as a regression?

noisysocks replied:
Yeah, so previously `TimePicker` rolled its own solution using regular `<input type="number">` elements and was written so that `onChange` would not be called if e.g. `value > max`. We're now using `NumberControl` though which is written so that `onChange` is called with `max` if `value > max`.

I don't really think either behaviour is better or worse, just that it should be consistent across components. I think letting `NumberControl` dictate what happens in this case is the correct approach as that's the point of using standardised components... to make these decisions for us!

ciampo replied:
Sounds good, I agree with your points!

ciampo commented on packages/components/src/date-time/time/test/index.tsx:
```diff
@@ -127,13 +154,18 @@ describe( 'TimePicker', () => {
+		await user.keyboard( '{Enter}' );
 
-		expect( onChangeSpy ).not.toHaveBeenCalled();
+		expect( onChangeSpy ).toHaveBeenCalledWith( '1986-10-18T11:59:00' );
```
Same as previous comment

ciampo commented on packages/components/src/date-time/date/styles.ts:
```diff
@@ -0,0 +1,58 @@
+
+export const Day = styled.div< { hasEvents: boolean } >`
+	align-content: center;
+	display: flex;
```
Out of curiosity, have you considered using the `Flex` (or `VStack`) components instead?

noisysocks replied:
It did not occur to me! I've swapped it out for `VStack`. This stuff is pretty new to me... do you think there's a benefit one way or the other? On the one hand it's nice that by using `display: flex` all the styling is together in `styles.ts`, but on the other hand it's nice that by using `VStack` we can remove a few lines of code.

ciampo replied:
The reason why I'm suggesting `VStack` is mostly because I believe that by being the first users of our own components, we can have more chances at improving the library and at promoting components composition

ciampo reviewed (commented):
Thank you @noisysocks for working on this, it's looking great!

-------

I've left most of my feedback via inline comments, but I'll add a few more points here:

> Re-organises DateTimePicker folder structure to be more consistent with newer components.
> Each sub-component in date-time now has its own folder.

Thank you for improving the folder structure! Ideally, to follow the latest [guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md), there would be more changes required (e.g. splitting each component into a `hook` and `component` files, adding READMES for each subcomponent...), but those changes can definitely wait and be applied at a later point as a follow-up.

Also, we normally keep only one `stories` folder per component family, instead of having a `stories` folder in each subcomponent's folder — would you mind applying those changes?

> Removes Reset button and Help functionality from DateTimePicker.
> The new design does not have this.
> This maybe is a breaking change. We can phase this in using a prop if we feel that it's too much of a change to impose on third parties.

For the sake of these changes, I would keep it around. We can look into adding a soft deprecation warning about it, and remove it later down the road (e.g. after WordPress 6.2 gets released) 

> I'll split out some of the @wordpress/components stuff into a separate PR for ease of review tomorrow, but don't let that stop you from giving this a whirl!

That would be awesome — I can see a few PRs can could be splintered off:

- updating `@types/react-dates`'s version
- updating `moment`'s version
- changes to `NumberControl` (if any)
- changes to `SelectControl`
- changes to `DateTimePicker`
- remaining updated in the `block-editor` and `edit-site` packages

But again — we can keep these changes together in the same PR for now, for the sake of iterating quicker and have a first few rounds of reviews

noisysocks commented:
Thanks for the reviews! I've addressed all the feedback. I'll split the PR up on Monday.

ciampo reviewed (commented):
Thank you! I left a few more replies to your comments

> Thanks for the reviews! I've addressed all the feedback. I'll split the PR up on Monday.

Awesome! I haven't had the time to look at all of your latest changes in details (like, for example, the opt-out flag for the reset button), but i'll make sure I do in each split PR

noisysocks commented:
OK I split these three PRs out:

- https://github.com/WordPress/gutenberg/pull/41266
- https://github.com/WordPress/gutenberg/pull/41269
- https://github.com/WordPress/gutenberg/pull/41270

Will switch this PR back to draft until they're merged.

I'd like to land everything else together in this PR as that way all of the user-facing changes are sure to appear at the same time.

ciampo commented:
> OK I split these three PRs out:

Thank you! All PRs have been approved :)

> I'd like to land everything else together in this PR as that way all of the user-facing changes are sure to appear at the same time.

Makes sense. My initial preference was that changes to `DateTimePicker` would be applied in a separate PR, but you make a valid point — let's proceed as you suggested

noisysocks committed:
Re-organise date-time dir to have component subdirs

noisysocks committed:
Update design of DateTimePicker and convert to Emotion

- Updates the design of DateTimePicker, DatePicker and TimePicker.
- Converts CSS in DateTimePicker, DatePicker and TimePicker to Emotion
  where it makes sense to do so.
- Allows removal of Reset button and Help functionality from
  DateTimePicker.
- Adds PublishDateTimePicker to block-editor. This exposes a
  specialised DateTimePicker that is good for choosing a post publish
  date.

noisysocks commented:
OK, I've rebased this to include those PRs above which were merged. This PR now contains just the user-facing changes.

I recommend setting GitHub to only show the second commit in this PR (b87608b) with _Hide whitespace_ checked so that the diff is less noisy.

noisysocks commented:
Some of the e2e failures are legitimate, probably because some classnames have been removed, I'll take a look at them tomorrow 🙂

ciampo commented on packages/components/src/date-time/date-time/index.tsx:
```diff
@@ -31,6 +32,8 @@ function UnforwardedDateTimePicker(
 		onMonthPreviewed = noop,
 		onChange,
 		events,
+		__nextRemoveHelpButton = false,
+		__nextRemoveResetButton = false,
```
We should probably add a deprecation warning for both props ([example](https://github.com/WordPress/gutenberg/blob/6eff9a514591f3b14f667e94b745eaaa507d912a/packages/components/src/card/card/hook.js#L26-L29))

ciampo commented on packages/components/src/date-time/date-time/index.tsx:
```diff
@@ -1,9 +1,6 @@
  */
-// Needed to initialise the default datepicker styles.
-// See: https://github.com/airbnb/react-dates#initialize
-import 'react-dates/initialize';
```
For ease of review, noting that this line has been moved to `packages/components/src/date-time/date/index.tsx`

ciampo commented on packages/components/src/date-time/time/styles.ts:
```diff
@@ -0,0 +1,119 @@
+export const MonthSelect = styled( SelectControl )`
+	height: 36px;
+
+	// Ideally would target the component instead of a classname.
+	.components-select-control__input {
+		line-height: 30px;
+	}
```
Have you tried using the `Select` styled component from `SelectControl` as the selector here? (excuses for the "selection" of terms used).

Something like

```diff
diff --git a/packages/components/src/date-time/time/styles.ts b/packages/components/src/date-time/time/styles.ts
index 7f8ff90de4..567f32c0c5 100644
--- a/packages/components/src/date-time/time/styles.ts
+++ b/packages/components/src/date-time/time/styles.ts
@@ -15,6 +15,7 @@ import {
 } from '../../input-control/styles/input-control-styles';
 import NumberControl from '../../number-control';
 import SelectControl from '../../select-control';
+import { Select } from '../../select-control/styles/select-control-styles';
 
 export const Wrapper = styled.div``;
 
@@ -96,8 +97,7 @@ export const MonthSelectWrapper = styled.div`
 export const MonthSelect = styled( SelectControl )`
 	height: 36px;
 
-	// Ideally would target the component instead of a classname.
-	.components-select-control__input {
+	${ Select } {
 		line-height: 30px;
 	}
 `;
```

(Haven't tested on my local machine though)

noisysocks replied:
That works, thanks! (Not sure why I didn't do this in the first place 😛)

ciampo commented on packages/components/src/date-time/time/styles.ts:
```diff
@@ -0,0 +1,119 @@
+export const DayInput = styled( NumberControl )`
+	${ baseInput }
+
+	width: 35px;
+`;
+
+export const YearInput = styled( NumberControl )`
+	${ baseInput }
+
+	width: 55px;
```
Are these numbers "pixel perfect" with the reference design? From what I remember, we usually try to stick to multiple of 4 (and use the `space` utility function)

noisysocks replied:
I think I copied them from from the old design. I'll look at using `space()`—good idea.

ciampo commented on packages/editor/src/components/post-schedule/index.js:
```diff
@@ -75,20 +74,14 @@ export default function PostSchedule() {
-	function onChange( newDate ) {
-		onUpdateDate( newDate );
-		const { ownerDocument } = ref.current;
-		ownerDocument.activeElement.blur();
```
Is the "blur" functionality being removed on purpose? (apologies in case I missed it)

noisysocks replied:
Yes this was in response to an earlier bit of feedback—see https://github.com/WordPress/gutenberg/pull/41097#issuecomment-1129646917.

ciampo replied:
Makes sense, thank you!

ciampo reviewed (commented):
> I recommend setting GitHub to only show the second commit in this PR with Hide whitespace checked so that the diff is less noisy.

Thank you for committing your changes in a reviewer-friendly way!

> Some of the e2e failures are legitimate, probably because some classnames have been removed, I'll take a look at them tomorrow 🙂

Good point. In the meantime I had another look at the changes and left a few more inline comments around.

noisysocks committed:
Add deprecated warning for when __nextRemoveHelpButton and __nextRemoveResetButton are not given

noisysocks committed:
Target component instead of classname

noisysocks committed:
Use space() instead of hardcoded widths

noisysocks committed:
Update CSS selectors used in E2E tests

noisysocks committed:
Update changelog

noisysocks committed:
Update scheduling e2e test

noisysocks commented:
OK, tests fixed! I want to keep the e2e changes minimal here so I did not switch away from using CSS selectors, even though that's not really ideal. It would be good to update `datepicker.test.js` to use Playwright / ARIA selectors in a future PR.

ciampo commented on packages/components/src/date-time/date-time/index.tsx:
```diff
@@ -37,6 +38,23 @@ function UnforwardedDateTimePicker(
 ) {
+	if ( ! __nextRemoveHelpButton ) {
+		deprecated( 'Help button in wp.components.DateTimePicker', {
+			since: '13.4',
+			version: '14.6', // Six months of plugin releases.
+			hint:
+				'Set the `__nextRemoveHelpButton` prop to `true` to remove this warning and opt in to the new behaviour, which will become the default in a future version.',
+		} );
+	}
+	if ( ! __nextRemoveResetButton ) {
+		deprecated( 'Reset button in wp.components.DateTimePicker', {
+			since: '13.4',
+			version: '14.6', // Six months of plugin releases.
```
Interesting. I believe that I always used the WordPress version (and not the Gutenberg plugin version) when working on components. Would be interesting to see which approach is the expected one!

noisysocks replied:
I've always seen and used the Gutenberg version... it's an interesting question because the code ends up both in the plugin and Core... hmmm! 🤔

andrewserong replied:
At least the version numbers in Gutenberg are _way_ ahead of core, so it's easy to tell them apart 😁

ciampo commented on packages/e2e-tests/specs/editor/various/datepicker.test.js:
```diff
@@ -15,12 +15,24 @@ async function getSelectedOptionLabel( selector ) {
 }
 
 async function getDatePickerValues() {
-	const year = await getInputValue( '[aria-label="Year"]' );
-	const month = await getInputValue( '[aria-label="Month"]' );
-	const monthLabel = await getSelectedOptionLabel( '[aria-label="Month"]' );
-	const day = await getInputValue( '[aria-label="Day"]' );
-	const hours = await getInputValue( '[aria-label="Hours"]' );
-	const minutes = await getInputValue( '[aria-label="Minutes"]' );
```
Just wondering if it wouldn't be a good thing to add these `aria-label`s in our code (unless those inputs are already labelled in an accessible way)

noisysocks replied:
The inputs are labelled, but it's done using `<label for="">` which you can't target using a CSS selector. I thought about using `puppeteer-testing-library` here so that we can do this properly but I think it's too great of a change for this PR.

ciampo replied:
Agreed!

ciampo reviewed (approved):
Apart from a couple of minor comments, code changes LGTM 🚀 

Given the size of this PR, I'd feel better if another pair of 👀 also took a look

andrewserong commented on packages/block-editor/src/components/publish-date-time-picker/README.md:
```diff
@@ -0,0 +1,54 @@
+`@wordpress/components` and adds additional post-specific controls.
+
+See [the documentation for
+DateTimePicker](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/date-time)
```
Tiny nit (and feel free to ignore): I think we can use an absolute path to the readme file to get links in the handbook to be directed to other pages within the handbook, rather than redirecting to the Gutenberg repo? E.g. links like the following: https://github.com/WordPress/gutenberg/blob/b964485122fa05ca305d544946ade1d3c1eeabc4/packages/components/CONTRIBUTING.md#L428

Ideally, for someone browsing in the handbook they'd ultimately be directed to: https://developer.wordpress.org/block-editor/reference-guides/components/date-time/

noisysocks replied:
Cool! I'll do that.

ramonjd commented on packages/block-editor/src/components/publish-date-time-picker/index.js:
```diff
@@ -0,0 +1,50 @@
+				<Button
+					className="block-editor-publish-date-time-picker__reset"
+					variant="tertiary"
+					onClick={ () => onChange?.( null ) }
```
Is there a reason why `onChange` is treated as optional but `onClose` not?

No biggie, just asking.

andrewserong replied:
From my reading of it, it looks like `onClose` is passed as a reference to the Button's `onClick`, so it can happily be `undefined`, whereas in this handler, `onChange` is being called, so it needs optional chaining in case it's `undefined`? I could be wrong, though!

noisysocks replied:
With `onChange` I'm just matching the behaviour of `DateTimePicker` which has it as optional.

You're right that `onClose` should probably also be optional for consistency. 

It just so happens that the next PR I have lined up (https://github.com/WordPress/gutenberg/pull/41362) makes `onClose` optional so let's avoid re-triggering GitHub Actions yet again and wait for that.

ramonjd replied:
Thanks for the explanation!

noisysocks committed:
Use relative URLs in README.md

andrewserong reviewed (approved):
Great work @noisysocks, this is feeling very polished to me while testing in the editor and Storybook, and the code change looks solid. Nice work moving the zero padding to a state reducer, too, that looks like a good and logical way to handle the padding behaviour for now.

LGTM! 🎉

ramonjd reviewed (approved):
This is working as described for me. I tested in the post editor and in storybook.

The label additions come through on MacOS Voice Over.

aaronrobertshaw reviewed (approved):
Nice work wrangling all of this @noisysocks 🙇 

✅ Storybook examples are functioning well
✅ Works great in the editor
✅ All unit tests and e2es are passing locally

LGTM! 🚢

ramonjd commented on packages/components/src/date-time/date/index.tsx:
```diff
@@ -15,12 +16,14 @@ import DayPickerSingleDateController from 'react-dates/lib/components/DayPickerS
+import { Day, NavPrevButton, NavNextButton } from './styles';
 
 const TIMEZONELESS_FORMAT = 'YYYY-MM-DDTHH:mm:ss';
 const ARIAL_LABEL_TIME_FORMAT = 'dddd, LL';
```
Not necessary for this PR (it's in trunk) I noticed that the aria-label on the day component outputs symbols.

<img width="1119" alt="Screen Shot 2022-05-27 at 11 17 15 am" src="https://user-images.githubusercontent.com/6458278/170611316-f0f499da-9e82-4916-8589-c285781077c7.png">


Not sure if the problem is in `moment()`. It's something to do with the locale aware format  `LL` 

This works

```js
const ARIAL_LABEL_TIME_FORMAT = 'dddd, MMMM DD, YYYY'; // Monday, May 09, 2022
```

But it's not locale aware I presume so it might need some further 👀

noisysocks replied:
Oh interesting, good catch. We probably should use our own `dateFormat` function that's in `@wordpress/date` here:

https://github.com/WordPress/gutenberg/blob/ef6c2fe106788c7efec627606b3de892974a0d9d/packages/components/src/date-time/date.tsx#L44

I'll turn your comment into an issue.

noisysocks replied:
https://github.com/WordPress/gutenberg/issues/41384

noisysocks commented:
Thanks everyone, big love ❤️
